File: packages/client/src/ffi/workspaces.ts
import { invoke } from "./invoke";

export type Workspace = {
  id?: number;
  uid: string;
  name: string;
  selected: boolean;
  created_at: string;
  updated_at: string;
  synced?: boolean;
};

type Args = {
  list_workspaces: never;
  create_workspace: {
    name: string;
    selected: boolean;
    uid?: string;
    created_at?: string;
    updated_at?: string;
    synced?: boolean;
  };
  delete_workspace: { uid: string };
  update_workspace: {
    uid: string;
    name: string;
    selected: boolean;
    synced?: boolean;
  };
  get_workspace_id: { name: string };
};

export const createWorkspace = async (
  name: string,
  selected: boolean,
  uid?: string,
  createdAt?: string,
  updatedAt?: string,
  synced?: boolean,
) => {
  try {
    const workspace = await invoke<Args, Workspace>()("create_workspace", {
      name,
      selected,
      uid,
      created_at: createdAt,
      updated_at: updatedAt,
      synced,
    });

    if (!workspace) {
      throw new Error("Empty response!");
      return;
    }

    return workspace;
  } catch (e) {
    console.error("[Workspace] Create failed!", e);
    throw e;
  }
};

export const getWorkspaces = async () => {
  try {
    const workspaces = await invoke<Args, Array<Workspace>>()(
      "list_workspaces",
      {} as never,
    );

    if (!workspaces) {
      throw "Unknown failure!";
    }

    return workspaces;
  } catch (e) {
    console.error("[Workspace] Failed to fetch!", e);
    throw e;
  }
};

export const deleteWorkspace = async (uid: string) => {
  try {
    if (
      !(await invoke<Args, boolean>()("delete_workspace", {
        uid,
      }))
    ) {
      throw new Error("Empty response!");
    }

    return true;
  } catch (e) {
    console.error("[Workspace] Failed to delete!", e);
    throw e;
  }
};

export const updateWorkspace = async (
  uid: string,
  name: string,
  selected: boolean,
  synced?: boolean,
) => {
  try {
    return await invoke<Args, void>()("update_workspace", {
      uid,
      name,
      selected,
      synced,
    });
  } catch (e) {
    console.error("[Workspace] Failed to update!", e);
    throw e;
  }
};

export const getByName = async (name: string) => {
  try {
    return await invoke<Args, number>()("get_workspace_id", {
      name,
    });
  } catch (e) {
    console.error("[Workspace] Failed to get by name!", e);
  }
};

File: crates/pyxis-app/src/handlers/workspaces.rs
use chrono::Utc;
use pyxis_shared::{database::Database, entities::workspaces::Workspace};
use tauri::State;

#[tauri::command]
pub fn create_workspace(
    name: String,
    selected: bool,
    created_at: Option<String>,
    updated_at: Option<String>,
    uid: Option<String>,
    synced: Option<bool>,
    database: State<Database>,
) -> Option<Workspace> {
    let workspace = Workspace::new(name, selected, None, created_at, updated_at, uid, synced);

    match workspace.create(&database.get_connection()) {
        Ok(_) => Some(workspace),
        Err(e) => {
            eprintln!("[Workspaces] Failed to create! {}", e);
            None
        }
    }
}

#[tauri::command]
pub fn list_workspaces(database: State<Database>) -> Option<Vec<Workspace>> {
    match Workspace::list(&database.get_connection()) {
        Ok(workspaces) => Some(workspaces),
        Err(e) => {
            eprintln!("[Workspaces] Failed to fetch! Error: {e}");
            None
        }
    }
}

#[tauri::command]
pub fn delete_workspace(uid: String, database: State<Database>) -> bool {
    match Workspace::delete(uid, &database.get_connection()) {
        Ok(_) => true,
        Err(e) => {
            eprintln!("[Workspaces] Failed to delete! {}", e);
            false
        }
    }
}

#[tauri::command]
pub fn update_workspace(
    uid: String,
    name: String,
    selected: bool,
    database: State<Database>,
    synced: Option<bool>,
) -> Option<Workspace> {
    let conn = &database.get_connection();
    let workspace = match Workspace::list(conn) {
        Ok(w) => w.into_iter().find(|w| w.uid == uid),
        Err(e) => {
            eprintln!("[Workspaces] Failed to get for update! {}", e);
            None
        }
    };

    if let Some(mut workspace) = workspace {
        workspace.name = name;
        workspace.selected = selected;
        workspace.updated_at = Utc::now().to_rfc3339();
        workspace.synced = synced;

        match workspace.update(conn) {
            Ok(_) => {
                return Some(workspace);
            }
            Err(e) => {
                eprintln!("[Workspaces] Failed to update! {}", e);
                return None;
            }
        }
    }

    None
}

#[tauri::command]
pub fn get_workspace_id(name: String, database: State<Database>) -> Option<i64> {
    match Workspace::get_by_name(&database.get_connection(), name) {
        Ok(workspace_id) => Some(workspace_id),
        Err(e) => {
            eprintln!("[Workspaces] Failed to get id! Error: {e}");
            None
        }
    }
}
